File: backend/app/schemas/exam.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime


class ExamQuestionCreate(BaseModel):
    subject: str
    year: Optional[int] = None
    question_type: str  # choice | fill | essay | reading
    content: str
    options: Optional[Dict[str, str]] = None
    answer: str
    explanation: Optional[str] = None
    difficulty: int = 3
    tags: List[str] = []


class ExamQuestionUpdate(BaseModel):
    subject: Optional[str] = None
    year: Optional[int] = None
    question_type: Optional[str] = None
    content: Optional[str] = None
    options: Optional[Dict[str, str]] = None
    answer: Optional[str] = None
    explanation: Optional[str] = None
    difficulty: Optional[int] = None
    tags: Optional[List[str]] = None


class ExamQuestionResponse(BaseModel):
    id: int
    subject: str
    year: Optional[int] = None
    question_type: str
    content: str
    options: Optional[Dict[str, str]] = None
    answer: str
    explanation: Optional[str] = None
    difficulty: int
    tags: List[str]
    created_at: datetime

    class Config:
        from_attributes = True


class ExamQuestionListResponse(BaseModel):
    items: List[ExamQuestionResponse]
    total: int
    page: int
    page_size: int
    pages: int


class VocabularyCreate(BaseModel):
    subject: str = "english"
    word: str
    phonetic: Optional[str] = None
    meaning: str
    example: Optional[str] = None
    level: Optional[str] = None


class VocabularyResponse(BaseModel):
    id: int
    subject: str
    word: str
    phonetic: Optional[str] = None
    meaning: str
    example: Optional[str] = None
    level: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class VocabularyListResponse(BaseModel):
    items: List[VocabularyResponse]
    total: int
    page: int
    page_size: int
    pages: int


class MockExamCreate(BaseModel):
    student_id: int
    title: str
    subject: str
    question_count: int = 20
    question_types: Optional[List[str]] = None
    difficulty_range: Optional[List[int]] = None
    tags: Optional[List[str]] = None


class MockExamResponse(BaseModel):
    id: int
    student_id: int
    title: str
    subject: str
    question_ids: List[int]
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    score: Optional[float] = None
    created_at: datetime
    questions: Optional[List[ExamQuestionResponse]] = None

    class Config:
        from_attributes = True

File: admin-web/src/api/exam.ts
import client from './client'
import type { PaginatedResponse } from '../types/api'

export interface ExamQuestion {
  id: number
  subject: string
  year: number
  question_type: string
  content: string
  options?: Record<string, string>
  answer: string
  explanation?: string
  difficulty: number
  tags: string[]
}

export interface VocabularyItem {
  id: number
  word: string
  phonetic?: string
  meaning: string
  example?: string
  level: string
}

export const examApi = {
  listQuestions: async (params?: {
    subject?: string
    year?: number
    question_type?: string
    difficulty?: number
    page?: number
    page_size?: number
  }): Promise<PaginatedResponse<ExamQuestion>> => {
    const response = await client.get<PaginatedResponse<ExamQuestion>>('/api/exam/questions', { params })
    return response.data
  },

  createQuestion: async (data: Omit<ExamQuestion, 'id'>): Promise<ExamQuestion> => {
    const response = await client.post<ExamQuestion>('/api/exam/questions', data)
    return response.data
  },

  listVocabulary: async (params?: {
    level?: string
    search?: string
    page?: number
    page_size?: number
  }): Promise<PaginatedResponse<VocabularyItem>> => {
    const response = await client.get<PaginatedResponse<VocabularyItem>>('/api/exam/vocabulary', { params })
    return response.data
  },
}
